### ISSUE TITLE
No uninitialized  warning for member variable and getter

### ISSUE THREAD

**Issue by fekir** [llvm/llvm-project#78129]
~~~~
#include <cstdio>

struct a{
    int i;
    int get_i() const {
        return i;
    }
    a() : i(get_i()){}
};


int main(){
    a instance = a();
    std::printf("%d", instance.i);
}
~~~~

clang does not generate any warning, even with optimizations enabled.

It generates a warning if `a() : i(get_i()){}` is changed to `a() : i(i){}`.

Example on godbolt, with comparison with GCC, which does generate a warning when optimizations are enabled

https://godbolt.org/z/6E43zroWj

**Comment by DimitryAndric** [comment]
I can at least confirm that clang trunk doesn't warn at all. So it created a nice "random" number generator:

```text
$ ./pr78129
548438528
$ ./pr78129
554314800
$ ./pr78129
543163808
$ ./pr78129
547619032
$ ./pr78129
539220120
$ ./pr78129
544666152
```

What is also interesting is that UBSan does not warn either:

```text
$ clang++ -O0 -g -fsanitize=undefined pr78129.cpp -o pr78129
$ ./pr78129
2354616
```

In any case I definitely agree that it _should_ warn. And UBSan should catch this somehow.

**Comment by dwblaikie** [comment]
Generally clang's warnings are not interprocedural - it's too expensive to do at compile time. So analyzing `get_i` to warn in the ctor is probably out of scope/infeasible.

UBSan catching this seems like something to look into, at least.
